src/helpers/map.rs

```diff
@@ -0,0 +1,31 @@
+/// Trait for transforming values in the manner of `Option::map` or `Iter::map`.
+///
+/// This is used to implement malicious downgrades.
+pub trait Map<M: Mapping> {
```

Comment:
I wish there was a way to express this without an extra generic type that has no meaning in `Map`. I guess associated type does not work because it is unstable?

Reply:
I think of the type parameter `M` as representing the degree of freedom to specify what mapping is being performed -- somewhat like the parameter `F` on [`iter::Map`](https://doc.rust-lang.org/std/iter/struct.Map.html).

On the impls of `Map<M>` for various types, the parameter signifies "this code is generic over the type of mapping being performed".

Associated type defaults are unstable, but I'm not sure what I would do with an associated type for `M`. I see how the type parameter looks like unnecessary baggage in most of the `Map` impls, but it's necessary at the leaf (`impl<F: Field> Map<UncheckedDowngrade> for AdditiveShare<F>`), and I'm not sure how to accomplish the same goal without it. 

This PR doesn't really demonstrate the power of the generic `Map`, but to illustrate, a version of `Map` that understood contexts and record IDs could replace the `Reshare` impls in [#493](https://github.com/private-attribution/ipa/pull/493/files#diff-eeea3aba623b98182cbcf841810cefca926225cbe6c0d766097b9b807831d577R227). Once `Map` is implemented once for `MCCappedCreditsWithAggregationBit`, you would automatically get `Reshare`, `Reveal`, `UpgradeMalicious`, `DowngradeMalicious`, and any other operation that can be applied to complex structs by applying the operation to each member of the struct.